### PR TITLE
[SR-13753][Diagnostics] Downgrade l-value unused diagnostics to warning

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
When expression type checks to a `l-value` no is produced `missing return` because lvalue unused is currently being diagnosed as an error and failing early on type checking, therefore not even getting into SIL dataflow diagnostics where missing return diagnostic would be produced. 
We can see this exactly on those two cases:
```swift

let _ : () -> Int = {
  let x : Int = 0 // x is a let so x would be an r-value, unused would be just a warning and correct SIL dataflow diagnostics would be produced. 
  var a : Int = 0
  
  x // error: missing return
}


let _ : () -> Int = {
  var x : Int = 0 // x is a var x result will typecheck as an l-value therefore fail early on typechecking 
  var a : Int = 0
  
  x // error: expression resolves to an unused variable
}
```
This just downgrades `l-value` to a warning and defer the proper diagnose errors in SIL. 

Resolves SR-13753.

### CODE REVIEW

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test

**Comment by varungandhi-apple** [comment]
I think this is correct, I don't see any reason why this should be an error and not a warning.

Not related to this PR but, I think the wording "expression resolves to an unused variable" seems unnecessarily jargon-heavy, there is no need to use the word "resolved". We may want to simplify the wording...

**Review by varungandhi-apple — APPROVED** [review]

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test macOS platform

**Comment by LucianoPAlmeida** [comment]
> I think the wording "expression resolves to an unused variable" seems unnecessarily jargon-heavy, there is no need to use the word "resolved". We may want to simplify the wording...

I agree, maybe just simply "result of expression is unused"? WDYT, I can put a follow-up ... or maybe we can create a starter bug on JIRA :)

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test macOS platform

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test macOS platform

**Comment by varungandhi-apple** [comment]
Creating a starter bug sounds good. 👍🏼

**Comment by LucianoPAlmeida** [comment]
> Creating a starter bug sounds good. 👍🏼

Awesome!

> For the suggested wording "result of expression is unused" -- I think it's a bit tricky, on one hand, I can understand what it is saying, but it's kind of funky because function application has a result, and expression evaluation has a result, but I'm not sure if the phrase might be confusing without the word "evaluation"... on the other hand, the word "evaluation" is again a bit of jargon...

I see ... so what about "result produced from expression is unused" or "result value produced from expression is unused" or "result value of the expression is unused"(I like this one)
